add a "restrictIds" function to pelias for debugging

blackmad opened the pull request:
The purpose of this is that often I want to know "why did document A
score above document B" - pulling open the entire ES debugging response
(especially when nested in the pelias response) makes it super hard to
keep track of what document I'm looking at.

With this change, internal links are added that ask pelias to run the
same query, just restricted to that document. Such that two of these can
be opened side by side for relatively easy comparison of scoring.

This adds links that look like 
![image](https://user-images.githubusercontent.com/445616/89786017-451d3a80-dae9-11ea-92cd-bdde174220e0.png)
http://localhost:3100/v1/search?text=52%20ten%20eyck%20st%20apt%203b%20brooklyn%20ny&debug=explain&api_key=ge-1e618cf3d7bd023d&restrictIds=openstreetmap%3Astreet%3Apolyline%3A12087878

@orangejulius we talked last week about how to make it easier / more useful to see the explain output for one specific result, here's my attempt at that.

I'm open to moving the query down to pelias-query if need be.

orangejulius commented:
Huh, first of all, I'm not sure we even knew about the [IDs query](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/query-dsl-ids-query.html). Maybe we didn't use it in the past because we used to much more heavily rely on the `_type` property before we upgraded to Elasticsearch 6/7. Pretty cool.

I both like the idea that this is very explicitly a "debug only" endpoint, and also wonder if there's a way we can make it generally useful. If not, then this PR is good to go. @missinglink do you have any thoughts?

orangejulius commented on query/view/restrict_ids.js:
```diff
@@ -0,0 +1,17 @@
+
+  return {
+    ids: {
+      type: '_doc',
```
Surprisingly, this parameter does not cause issues on ES7, despite the [ES7.8 docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/query-dsl-ids-query.html) not listing the `type` parameter.

In general, I think most `type` related functionality was deprecated in ES6 and completely removed in ES7, but I guess this one is silently ignored for now.

I just opened a ticket to discuss [dropping support for ES6](https://github.com/pelias/pelias/issues/881). If we do drop ES6 support, we should just remove this line. If not, we should _technically_ use the value of `schema.typeName` from [pelias-config](https://github.com/pelias/config/blob/master/config/defaults.json#L73)

blackmad replied:
oh man this took me so long to figure out you meant elasticsearch and not ecmascript.

will plumb schema.typeName through for now?

blackmad replied:
I take that back, plumbing through schema.typeName will be a bit gross. I'd rather drop this field

blackmad commented:
re: limitations of this - if you put an id in here that is filtered out by the other parts of the filter query (like parentids) the query returns nothing.

blackmad commented:
bump

blackmad commented:
maybe make parameter debug.restrictIds?

blackmad commented:
do you want the name of this field changed?

orangejulius commented:
So I happened to take a look at this PR yesterday, hoping it would help me debug why a particular record was not showing up in a query when I thought it should. As a result I learned that this PR didn't do what I thought it did.

What I _thought_ was that the `restrictIds` parameter would modify any actual Elasticsearch queries generated  and run by Pelias so that they filtered out all records except the one(s?) listed. But it looks like instead it prints a debug entry with a URL to a separate query that can be executed directly against an Elasticsearch instance that filters out all records except the one(s) listed.

This doesn't seem too useful to me for two reasons:
1. Having to execute that query manually against Elasticsearch is pretty time consuming (in many cases the URL itself won't be valid if Elasticsearch is running on a separate instance or server, etc)
2. The URL attempts to encode an entire Elasticsearch query in the URL, which I don't think is going to work very often. In my case it was a `structured` endpoint query and it exceeded the 4096 character limit for Elasticsearch queries in URLs, so all I got was an error.

Now that I know what this PR currently does, I don't think we really need it as is. To debug why a _particular_ result does not come up in a query there are two useful techniques that already work:
- filter to a small set of documents with a given parent with `boundary.gid`
- Filter to a small geographic area with `boundary.rect` or `boundary.circle` (maybe we make it easier to generate queries that target _really small_ areas around a particular record with one of these filters using the compare app or another frontend tool)

If we modified this PR to actually change the Elasticsearch queries executed as part of the Pelias query, so that it only record with the given GID(s) could be returned, then I think it would be useful. But even then I'm not sure it's worth adding a new parameter and maintaining it for all time, when the two techniques above already work pretty well with parameters that are more generally useful.

Do you want to modify this PR as I described? Or is it better to build up tools to use existing parameters to help debugging?

blackmad commented:
I really thought I had modified this to modify the query pelias runs for
the reasons discussed. Maybe something got lost. I’ll edit this on Monday.

On Fri, Oct 2, 2020 at 1:28 PM Julian Simioni <notifications@github.com>
wrote:

>
>
> So I happened to take a look at this PR yesterday, hoping it would help me
> debug why a particular record was not showing up in a query when I thought
> it should. As a result I learned that this PR didn't do what I thought it
> did.
>
>
> What I *thought* was that the restrictIds parameter would modify any
> actual Elasticsearch queries generated and run by Pelias so that they
> filtered out all records except the one(s?) listed. But it looks like
> instead it prints a debug entry with a URL to a separate query that can be
> executed directly against an Elasticsearch instance that filters out all
> records except the one(s) listed.
>
>
> This doesn't seem too useful to me for two reasons:
>
>
>
>
>
>    1. Having to execute that query manually against Elasticsearch is
>    pretty time consuming (in many cases the URL itself won't be valid if
>    Elasticsearch is running on a separate instance or server, etc)
>
>    2. The URL attempts to encode an entire Elasticsearch query in the
>    URL, which I don't think is going to work very often. In my case it was a
>    structured endpoint query and it exceeded the 4096 character limit for
>    Elasticsearch queries in URLs, so all I got was an error.
>
>
>
>
> Now that I know what this PR currently does, I don't think we really need
> it as is. To debug why a *particular* result does not come up in a query
> there are two useful techniques that already work:
>
>
>
>
>
>    - filter to a small set of documents with a given parent with
>    boundary.gid
>
>    - Filter to a small geographic area with boundary.rect or
>    boundary.circle (maybe we make it easier to generate queries that
>    target *really small* areas around a particular record with one of
>    these filters using the compare app or another frontend tool)
>
>
>
>
> If we modified this PR to actually change the Elasticsearch queries
> executed as part of the Pelias query, so that it only record with the given
> GID(s) could be returned, then I think it would be useful. But even then
> I'm not sure it's worth adding a new parameter and maintaining it for all
> time, when the two techniques above already work pretty well with
> parameters that are more generally useful.
>
>
> Do you want to modify this PR as I described? Or is it better to build up
> tools to use existing parameters to help debugging?
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pelias/api/pull/1482#issuecomment-702860763>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADMZMEMWHQOHHQIVZJWHCDSIYETJANCNFSM4PZ6FWFA>
> .
>
>
> --
*David Blackman*
creative technologist & wandering
help me find my purpose <http://purpose.blackmad.com>

blackmad commented:
Coming back to this PR after a bit

- this does do what you expected, just not for structured geocoding, I can add it (it does for autocomplete and search)

- I've found the ES links super useful since I can just click straight to them and it makes it super easy to have two parallel queries up on my screen. I know in your configuration this doesn't work because your ES hosts I think are even more VPN'd away? I haven't run into trouble with the query length, surprised that 4096 wasn't enough for a structured query. I can delete this link logic if you want though.
---> I could instead or additionally make these links back to the pelias host with the restrictId added?

- modifying the query with boundary radius games isn't something I love since it changes the query and the whole point is I'm trying to understand exactly what happened in a given run. It's also kind of awkward since the parameter names are slightly different for point vs circle

blackmad commented:
also there was talk of maybe changing this to be named debug.restrictIds, though that might be a little odd since there's already a top level param called "debug" if we think of these dotted cgi params as flattened json dicts
